### PR TITLE
perf(qwen3): drop the duplicate cublasLt tuning burst from startup

### LINKED ISSUE #748: qwen3: startup runs the cublasLt tuning sweep twice
## Description

On a single sm_89 GPU with Qwen3-4B and a warm page cache, HTTP-ready takes about 5.7 s, and instrumenting the engine start shows most of it is not weight upload:

| phase | time |
|---|---|
| weight upload (mmap -> GPU) | ~1.7 s |
| cublasLt tune, memory-profile thread | ~1.6 s |
| cublasLt tune, serving worker bind | ~1.5 s |
| munmap of the weight mapping | ~0.15 s |
| memory profile allocs + dummy step | ~0.1 s |
| KV pool + executor init | <0.1 s |

The tuning sweep (`gemm_lt_tune_cuda`) benches up to 16 heuristic algos for 23 iterations per (M, N, K), for every projection group and decode batch bucket. The tuned-plan cache is thread_local, and startup runs the sweep on two threads: once on the memory-profile worker (`profile_kv_budget_on_worker`) and again when the serving worker binds (`LocalQwen3Lane::bind`). The winner depends on the device and shape, not the thread, so the second sweep re-derives the same information — and since selection is timing-noise sensitive, the two threads can even end up on different algos for the same shape.

## Possible directions:

- Publish tuned winners in a process-level (device, M, N, K) map so later threads build their thread-local plans from the recorded algo without re-benchmarking. This also makes Tuned-policy selection consistent across threads within one process.
- Move the multi-gigabyte munmap off the startup critical path; nothing references the mapping once every weight is uploaded.

The Pin policy path is unaffected either way (it returns before the sweep).

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Fixes #748 

Startup runs the cublasLt decode-GEMM tuning sweep twice: once on the memory-profile thread and again when the serving worker binds, because the tuned-plan cache is thread_local. The sweep benches up to 16 algos for 23 iterations per (M, N, K), which costs about 1.5 s per thread on the measured host. The winner only depends on the device, not the thread, so this adds a process-level single-flight (device, M, N, K) -> algo map: the first thread to claim a shape benchmarks and publishes it, and any concurrent thread waits on the claim and adopts the published winner instead of re-benchmarking (a failed owner erases its claim so a waiter retries). The Pin policy path is untouched (it returns before the sweep), and Tuned-policy algo selection becomes consistent across threads in one process instead of two independently benchmarked, potentially divergent picks. The second commit moves the multi-gigabyte safetensors munmap off the startup critical path — nothing references the mapping once every weight is uploaded — and fails the load outright if the unmap thread cannot be spawned.

## Test Env

Measured on one sm_89 GPU, Qwen3-4B, warm page cache, three repetitions with medians: the serving worker's tune drops 1491 ms -> 0 ms, engine load 5160 -> 3663 ms, HTTP-ready 5743 -> 4099 ms (-29%).

Validation on the same GPU: `hf_golden_gate` (the serving thread now runs reused algos; logits still match the HF golden), `sampling_behavior`, `scheduler_robustness`, `prefix_cache`, `cached_tokens_usage`, both context-window tests, and the crate unit tests all pass; fmt and clippy clean.

## Type of Change

- [x] Optimisation (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project (see `docs/conventions/coding-style.md`).
- [x] I have performed a self-review of my own code.
- [x] I have formatted my commits according to Commitizen conventions.
- [x] I have run the local test suite and all tests pass (see `CLAUDE.md`).

### CODE REVIEW

File: openinfer-kernels/csrc/shared/linear.cu [hunk 1]
```diff
@@ -435,6 +447,48 @@ int gemm_lt_tune_cuda(const __nv_bfloat16 *const *Ws, int num_ws, int M, int N,
...
+  const std::array<int, 4> shared_key{device, M, N, K};
+  {
+    std::unique_lock<std::mutex> lock(g_lt_tuned_mu);
+    for (;;) {
+      auto it = g_lt_tuned_algos.find(shared_key);
+      if (it == g_lt_tuned_algos.end()) {
+        g_lt_tuned_algos.emplace(shared_key, LtTunedEntry{});
+        break;
+      }
+      if (it->second.ready) {
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply openinfer-kernels/csrc/shared/linear.cu:464]: Thanks. The call-sequence observation is correct, but I don't think it establishes a CUDA Graph capture violation.

The profile and serving workers bind the same model CUDA context. The profile sweep executes every candidate, including the published winner, before the serving worker is created. Destroying the profile thread's local descriptors, handle, and workspace does not destroy that CUDA context or unload the kernel.

cuBLAS routines can generally be captured, that a cuBLASLt handle is not tied to a particular CUDA context, and that cublasLtMatmulAlgo_t can be serialized and reused with the same cuBLAS version. The serving worker also supplies its own explicit 32 MiB workspace.

The hf_golden_gate exercised this adopted-plan path through batched CUDA Graph capture/replay and passed on the tested sm_89/cuBLAS environment. This differs from the Pin-path concern: Pin selects a heuristic without executing that algorithm anywhere, whereas the Tuned winner here has already run in the same CUDA context.

Without a reproduction or an API requirement that the algorithm must execute once per handle, I don't think this is a merge blocker.

File: openinfer-qwen3/src/weights.rs [hunk 2]
```diff
@@ -615,7 +615,12 @@ impl Qwen3Model {
             t_gpu.elapsed().as_secs_f64() * 1e3
         );
         drop(shards);
-        drop(mmaps);
+        // Page-table teardown of the multi-GB mapping is not worth blocking
+        // startup on; nothing references the mmaps once every weight is uploaded.
+        std::thread::Builder::new()
+            .name("qwen3-weights-unmap".into())
+            .spawn(move || drop(mmaps))
+            .map_err(|e| anyhow::anyhow!("weight unmap thread spawn failed: {e}"))?;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply openinfer-qwen3/src/weights.rs:623]: Thanks. The failure behavior here is intentional.

We considered both ignoring the spawn error and falling back to synchronous unmapping, but we follows fail-loud semantics and does not silently disable an optimization under resource pressure.

If spawning fails, the unstarted closure is dropped, so its captured mmaps are also dropped; there is no mapping leak. The spawn failure is then propagated with explicit context. Continuing with a synchronous drop would preserve functional correctness, but it would silently put the multi-GB munmap back onto the startup critical path.

For that reason, we are keeping the explicit load failure rather than adding a synchronous fallback.
